src/mongocxx/lib/mongocxx/v1/write_concern.cpp

```diff
@@ -12,4 +12,233 @@
+write_concern& write_concern::tag(bsoncxx::v1::stdx::string_view v) {
+    libmongoc::write_concern_set_wtag(to_mongoc(_impl), std::string{v}.c_str());
+    return *this;
+}
+
+bsoncxx::v1::stdx::optional<bsoncxx::v1::stdx::string_view> write_concern::tag() const {
+    bsoncxx::v1::stdx::optional<bsoncxx::v1::stdx::string_view> ret;
+
+    if (auto const str = libmongoc::write_concern_get_wtag(to_mongoc(_impl))) {
+        ret.emplace(str);
+    }
+
+    return ret;
+}
+
+write_concern& write_concern::journal(bool j) {
+    libmongoc::write_concern_set_journal(to_mongoc(_impl), j);
+    return *this;
+}
+
+bsoncxx::v1::stdx::optional<bool> write_concern::journal() const {
+    bsoncxx::v1::stdx::optional<bool> ret;
+
+    if (libmongoc::write_concern_journal_is_set(to_mongoc(_impl))) {
+        ret.emplace(libmongoc::write_concern_get_journal(to_mongoc(_impl)));
+    }
+
+    return ret;
+}
+
+bool write_concern::is_acknowledged() const {
+    return libmongoc::write_concern_is_acknowledged(to_mongoc(_impl));
+}
+
+bsoncxx::v1::document::value write_concern::to_document() const {
+    scoped_bson doc;
+
+    switch (this->acknowledge_level()) {
+        case level::k_default:
+            break;
+
+        case level::k_majority:
+            doc += scoped_bson{BCON_NEW("w", BCON_UTF8("majority"))};
+            break;
+
+        case level::k_tag:
+            doc += scoped_bson{BCON_NEW("w", BCON_UTF8(libmongoc::write_concern_get_wtag(to_mongoc(_impl))))};
+            break;
+
+        case level::k_unacknowledged:
+            doc += scoped_bson{BCON_NEW("w", BCON_INT32(0))};
+            break;
+
+        case level::k_acknowledged:
+            doc += scoped_bson{BCON_NEW("w", BCON_INT32(*this->nodes()))};
+            break;
+
+        case level::k_unknown:
+            // Precondition violation: undocumented but well-defined behavior.
+            break;
+    }
+
+    if (auto const opt = this->journal()) {
+        doc += scoped_bson{BCON_NEW("j", BCON_BOOL(*opt))};
+    }
+
+    {
+        auto const t = this->timeout().count();
+
+        if (t > 0) {
+            // Backward compatibility: only use "$numberLong" when necessary.
+            if (t <= std::int64_t{INT32_MAX}) {
+                doc += scoped_bson{BCON_NEW("wtimeout", BCON_INT32(static_cast<std::int32_t>(t)))};
+            } else {
+                doc += scoped_bson{BCON_NEW("wtimeout", BCON_INT64(t))};
+            }
+        }
+    }
+
+    return std::move(doc).value();
+}
+
+bool operator==(write_concern const& lhs, write_concern const& rhs) {
+    // Lexicographic comparison of data members.
+    return std::make_tuple(lhs.acknowledge_level(), lhs.timeout(), lhs.nodes(), lhs.tag(), lhs.journal()) ==
+           std::make_tuple(rhs.acknowledge_level(), rhs.timeout(), rhs.nodes(), rhs.tag(), rhs.journal());
+}
+
+write_concern::write_concern(void* impl) : _impl{impl} {}
+
+write_concern write_concern::internal::make(mongoc_write_concern_t* rc) {
+    return {rc};
```

Comment:
```suggestion
    return write_concern(static_cast<void*>(rc));
```

Suggest using an explicit cast to make it clear which constructor is being called. This comment also assumes you've applied the the suggestion from [my other comment](https://github.com/mongodb/mongo-cxx-driver/pull/1504/files#r2550784031). If not, I would still prefer to see `{static_cast<void*>(rc)}`.

Reply:
I am not aware of a diagnostic (either by the compiler or by clang-tidy) which warns for implicit cast to `void*` (ignoring the ones concerning integer <-> pointer casts). If there was one, I might be more inclined towards adopting this suggestion, but AFAICT there is no programmatic way to detect/enforce a "conversion to `void*` must be explicit" practice.